### PR TITLE
`Offline Entitlements`: integration tests

### PR DESCRIPTION
**Pull request by NachoSoto**
## Example tests:
```swift
func testOfferingsAreCachedInMemory() async throws {
    let onlineOfferings = try await Purchases.shared.offerings()
    expect(onlineOfferings.all).toNot(beEmpty())

    self.serverDown()

    let offlineOfferings = try await Purchases.shared.offerings()
    expect(offlineOfferings) === onlineOfferings
}

func testOfferingsAreCachedOnDisk() async throws {
    let onlineOfferings = try await Purchases.shared.offerings()
    expect(onlineOfferings.all).toNot(beEmpty())

    self.serverDown()
    await self.resetPurchases()

    let offlineOfferings = try await Purchases.shared.offerings()
    expect(offlineOfferings.response) == onlineOfferings.response

    let offering = try XCTUnwrap(offlineOfferings.current)
    expect(offering.availablePackages.count) == onlineOfferings.current?.availablePackages.count
    expect(offering.monthly?.storeProduct.productIdentifier) == "com.revenuecat.monthly_4.99.1_week_intro"
}
```

### CODE REVIEW

File: Sources/Misc/DangerousSettings.swift [hunk 1]
```diff
@@ -100,3 +97,16 @@ import Foundation
 }
 
 extension DangerousSettings: Sendable {}
+
+/// Dangerous settings not exposed outside of the SDK.
+internal protocol InternalDangerousSettingsType: Sendable {
```
Comment (NachoSoto) [inline Sources/Misc/DangerousSettings.swift:102]: This allows having a custom type in integration tests so we can swap the return value, without having to make the actual `DangerousSettings.Internal` `struct` mutable.

File: Sources/Misc/DangerousSettings.swift [hunk 2]
```diff
@@ -13,14 +13,11 @@ import Foundation
  */
 @objc(RCDangerousSettings) public final class DangerousSettings: NSObject {
 
-    /// Dangerous settings not exposed outside of the SDK.
-    internal struct InternalSettings {
+    internal struct Internal: InternalDangerousSettingsType {
```
Comment (NachoSoto) [inline Sources/Misc/DangerousSettings.swift:16]: Changed this from `DangerousSettings.InternalSettings` to `DangerousSettings.Internal`.

File: Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift [hunk 3]
```diff
@@ -26,7 +26,7 @@ class CustomerInfoResponseHandler {
         purchasedProductsFetcher: PurchasedProductsFetcherType,
         productEntitlementMapping: ProductEntitlementMapping?,
         userID: String,
-        offlineEntitlementsEnabled: Bool = false
+        offlineEntitlementsEnabled: Bool = true
```
Comment (NachoSoto) [inline Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift:29]: ⚠️  This PR enables offline entitlements (so we can test them on integration tests).
I think instead I'll get rid of the flag?

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+        self.serverDown()
+        try await self.purchaseMonthlyProduct()
+
+        logger.verifyMessageWasLogged(Strings.offlineEntitlements.computing_offline_customer_info, level: .info)
+        logger.verifyMessageWasNotLogged("Finishing transaction")
+    }
+
+    @available(iOS 15.2, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testPurchaseWhileServerIsDownPostsReceiptAfterServerComesBack() async throws {
```
Comment (NachoSoto) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:98]: I'm glad I'm running these in both SK1 and SK2. Turns out this is failing in SK1 only.

Comment (NachoSoto) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:98]: For some reason, after restarting the SDK and recreating `SKPaymentQueue`, the SDK isn't getting notified about the transaction again, unlike with SK2. Need to figure that out.

File: Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift [hunk 5]
```diff
@@ -86,6 +88,14 @@ class BaseBackendIntegrationTests: XCTestCase {
         self.mainThreadMonitor = nil
     }
 
+    /// Simulates closing the app and re-opening with a fresh instance of `Purchases`.
+    final func resetPurchases() async {
```
Comment (winstondu) [inline Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift:92]: Nit: naming could be a bit better, as it could be semantically confused with resetting "purchases of entitlements" rather than the "Purchases Singleton"

Comment (NachoSoto) [reply Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift:92]: I agree 👍🏻 I think I'll rename to `resetSingleton`

File: Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift [hunk 6]
```diff
@@ -156,7 +172,14 @@ private extension BaseBackendIntegrationTests {
...
-                     internalSettings: .init(enableReceiptFetchRetry: true))
+                     internalSettings: self)
     }
 
 }
+
+extension BaseBackendIntegrationTests: InternalDangerousSettingsType {
+
+    var enableReceiptFetchRetry: Bool { return true }
+    var forceServerErrors: Bool { return false }
```
Comment (winstondu) [inline Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift:183]: I think composition over inheritance is the way to go here.

It feels unnatural for the test class itself, which is fairly large and complex, to conform to this protocol and being what gets "dependency-injected" into the `DangerousSettings` initializer.

Why not have `BaseBackendIntegrationTests` instances hold a member variable of a new type, say a `MutableInternalDangerousSettings` class, that conforms to this protocol? 

The `serverUp()` and `serverDown()` methods can then just mutate `forceServerErrors` that variable.

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    var serverIsDown: Bool = false
+    override var forceServerErrors: Bool { return self.serverIsDown }
+
+    override func setUp() async throws {
+        try await super.setUp()
+
+        await self.waitForPendingCustomerInfoRequests()
+        if #available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *) {
+            try await self.ensureEntitlementMappingIsAvailable()
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:35]: Quick question, I guess `serverIsDown` is reset to `false` automatically on each test right?

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        self.serverDown()
+        await self.resetPurchases()
+
+        let offlineOfferings = try await Purchases.shared.offerings()
+        expect(offlineOfferings.response) == onlineOfferings.response
+
+        let offering = try XCTUnwrap(offlineOfferings.current)
+        expect(offering.availablePackages.count) == onlineOfferings.current?.availablePackages.count
+        expect(offering.monthly?.storeProduct.productIdentifier) == "com.revenuecat.monthly_4.99.1_week_intro"
+    }
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:62]: I think these offerings tests are not directly related to offline entitlements so not sure if they should be separate... But this is fine for now.

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testOfflineCustomerInfoWithNoPurchases() async throws {
+        Purchases.shared.invalidateCustomerInfoCache()
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:66]: Not directly related to this PR but looking at this comment https://github.com/RevenueCat/purchases-ios/blob/main/Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift#L144-L154. Looks like that waits for get offerings to finish, but not for the get customer info request to finish I think? If so, there could possibly be race conditions of getting the result for that request after this line is executed... There might be something I'm missing though!

**Review by tonidero — COMMENTED** [review]
Really love these tests! I think I should be able to copy part of it for the Android tests 👍. Though we don't support coroutines there yet so they won't look nearly as nice 😬

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift (hunk 9 above)
Comment (tonidero) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:66]: Nvm! I didn't see `waitForPendingCustomerInfoRequests`

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift (hunk 7 above)
Comment (NachoSoto) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:35]: Oops good catch, I don't think it was!

Comment (NachoSoto) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:37]: Done.

File: Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift (hunk 6 above)
Comment (NachoSoto) [reply Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift:183]: > I think composition over inheritance is the way to go

Normally I agree, but I fail to see how that would improve this particular case. That would semantically be the same.
This set up is much simpler this way. The way these tests are set up is such that the tests themselves become the configuration, which allows us to override flags like [this](https://github.com/RevenueCat/purchases-ios/blob/662940100f0cd1f4005934d2a99b4ba16e9ea6d3/Tests/BackendIntegrationTests/StoreKitObserverModeIntegrationTests.swift#L25).

What benefit does your suggestion provide?

**Comment by NachoSoto** [comment]
> Though we don't support coroutines there yet so they won't look nearly as nice 

@tonidero migrating tests to async (#1332) was huge, it allowed us to iterate, maintain, and write new integration tests much faster. Before it was callback hell.
Might be worth investing on that in iOS.

File: Tests/UnitTests/TestHelpers/AsyncTestHelpers.swift [hunk 10]
```diff
@@ -90,4 +90,13 @@ func asyncWait(
         line: line,
         foundCorrectValue
     ).to(beTrue(), description: description)
+
+    if !foundCorrectValue {
+        struct ConditionFailedError: Error {}
+
+        // Because this method is `async`, for some reason Swift is continuing execution of the test
```
Comment (NachoSoto) [inline Tests/UnitTests/TestHelpers/AsyncTestHelpers.swift:97]: This was required to avoid failing tests from crashing instead of failing.

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift (hunk 8 above)
Comment (NachoSoto) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:84]: It's related to being able to purchase while offline :P

File: Sources/Purchasing/StoreKit1/StoreKit1Wrapper.swift [hunk 11]
```diff
@@ -112,6 +114,34 @@ class StoreKit1Wrapper: NSObject {
         return payment
     }
 
+    private func notifyDelegateOfExistingTransactions() {
```
Comment (tonidero) [inline Sources/Purchasing/StoreKit1/StoreKit1Wrapper.swift:117]: Should we add something to the name of this method to indicate that it's only going to be run on tests? Like `notifyDelegateOfExistingTransactionsInTests`?

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        // 2. "Re-open" the app
+        await self.resetSingleton()
+
+        // 3. `CustomerInfo` should contain offline purchase
+        let info = try await Purchases.shared.customerInfo()
+        try await self.verifyEntitlementWentThrough(info)
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testPurchaseAgainAfterServerRecovers() async throws {
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:187]: I believe apple/Google shouldn't allow you to purchase the same subscription a second time right? So not sure if this is a valid scenario.

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+class OfflineWithNoMappingStoreKitIntegrationTests: BaseOfflineStoreKitIntegrationTests {
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testOfflineCustomerInfoFailsIfNoEntitlementMapping() async throws {
+        self.serverDown()
+
+        do {
+            _ = try await Purchases.shared.customerInfo(fetchPolicy: .fetchCurrent)
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:249]: Hmm so this will fail but not try to enter offline entitlements mode because it already has a cached customer info. I think we are missing a `invalidateCustomerInfoCache` before this?

**Review by tonidero — APPROVED** [review]
Looks great! This brought a great point in favor of supporting coroutines in Android 🙏

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+        Purchases.shared.invalidateCustomerInfoCache()
+        await self.resetSingleton()
+
+        // 6. To ensure (with a clean cache) that the receipt was posted
+        let info = try await Purchases.shared.customerInfo()
+        try await self.verifyEntitlementWentThrough(info)
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testReopeningAppWithOfflineEntitlementsDoesNotReturnStaleCache() async throws {
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:173]: Should we add a test case for foregrounding the app sends unsynced purchases?

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+    override func setUp() async throws {
+        try await super.setUp()
+
+        if #available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *) {
+            try await self.ensureEntitlementMappingIsAvailable()
+        }
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testOfflineCustomerInfoFailsIfNoEntitlementMapping() async throws {
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:52]: Oh wait, I think this test should be removed since it's in its own class

File: Sources/Purchasing/StoreKit1/StoreKit1Wrapper.swift (hunk 11 above)
Comment (NachoSoto) [reply Sources/Purchasing/StoreKit1/StoreKit1Wrapper.swift:117]: Good idea. Maybe `notifyDelegateOfExistingTransactionsIfNeeded`

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift (hunk 13 above)
Comment (NachoSoto) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:249]: Oh yeah, this is failing because it's not even trying to compute it because there's something in the cache 👍🏻 
Great catch! I'll add that.

File: Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift (hunk 15 above)
Comment (NachoSoto) [reply Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift:52]: Thanks for catching it.
